### PR TITLE
Feature/chatglm

### PR DESCRIPTION
**Pull request by klhhhhh**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: =2.0 [hunk 1]
```diff
@@ -0,0 +1,134 @@
+Defaulting to user installation because normal site-packages is not writeable
+Collecting protobuf
+  Using cached protobuf-4.23.4-cp37-abi3-manylinux2014_x86_64.whl (304 kB)
+Requirement already satisfied: transformers==4.30.2 in /home/lclk/.local/lib/python3.9/site-packages (4.30.2)
+Collecting cpm_kernels
```
Comment (FrankLeeeee) [inline =2.0:5]: Remove this file please.

File: colossalai/shardformer/policies/autopolicy.py [hunk 2]
```diff
@@ -104,6 +104,9 @@ class PolicyLocation:
     # Sam
     "transformers.models.sam.modeling_sam.SamModel":
         PolicyLocation(file_name="sam", class_name="SamModelPolicy"),
+    # ChatGLM
+    "tests.kit.model_zoo.transformers.chatglm2_6b.modeling_chatglm.ChatGLMModel":
```
Comment (FrankLeeeee) [inline colossalai/shardformer/policies/autopolicy.py:108]: You should not register like this, you should register with the normal class name when using chatglm with Transformers.

File: tests/kit/model_zoo/transformers/chatglm.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+# ================================
+
+config = ChatGLMConfig(num_layers=1, padded_vocab_size=65024, hidden_size=64, num_attention_heads=8, rmsnorm=False)
+
+config.__setattr__('original_rope', True)
+config.__setattr__('use_cache', True)
+
+
+def data_gen():
+
```
Comment (FrankLeeeee) [inline tests/kit/model_zoo/transformers/chatglm.py:19]: Remove this blank line.

File: tests/kit/model_zoo/transformers/chatglm.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+from .chatglm2_6b.modeling_chatglm import ChatGLMModel
+
+# ================================
+# Register single-sentence ChatGLM
+# ================================
+
+config = ChatGLMConfig(num_layers=1, padded_vocab_size=65024, hidden_size=64, num_attention_heads=8, rmsnorm=False)
+
+config.__setattr__('original_rope', True)
+config.__setattr__('use_cache', True)
```
Comment (FrankLeeeee) [inline tests/kit/model_zoo/transformers/chatglm.py:15]: Why not pass it to the `ChatGLMConfig` directly?

File: tests/kit/model_zoo/transformers/chatglm2_6b/MODEL_LICENSE [hunk 5]
```diff
@@ -0,0 +1,33 @@
+The ChatGLM2-6B License
```
Comment (FrankLeeeee) [inline tests/kit/model_zoo/transformers/chatglm2_6b/MODEL_LICENSE:1]: You don't need to put the license here, you can just mention this license at the top of the `modeling_chatglm.py`.

File: tests/kit/model_zoo/transformers/chatglm2_6b/MODEL_LICENSE [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+5. Limitation of Liability
+
+EXCEPT TO THE EXTENT PROHIBITED BY APPLICABLE LAW, IN NO EVENT AND UNDER NO LEGAL THEORY, WHETHER BASED IN TORT, NEGLIGENCE, CONTRACT, LIABILITY, OR OTHERWISE WILL ANY LICENSOR BE LIABLE TO YOU FOR ANY DIRECT, INDIRECT, SPECIAL, INCIDENTAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES, OR ANY OTHER COMMERCIAL LOSSES, EVEN IF THE LICENSOR HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
+
+6. Dispute Resolution
+
+This license shall be governed and construed in accordance with the laws of People’s Republic of China. Any dispute arising from or in connection with this License shall be submitted to Haidian District People's Court in Beijing.
+
+Note that the license is subject to update to a more comprehensive version.  For any questions related to the license and copyright, please contact us at glm-130b@googlegroups.com.
```
Comment (FrankLeeeee) [inline tests/kit/model_zoo/transformers/chatglm2_6b/MODEL_LICENSE:33]: Please use pre-commit to format your code.

File: tests/kit/model_zoo/transformers/chatglm2_6b/quantization.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
+from torch.nn import Linear
+from torch.nn.parameter import Parameter
+
```
Comment (FrankLeeeee) [inline tests/kit/model_zoo/transformers/chatglm2_6b/quantization.py:3]: This file is not needed if no quantization is tested.

File: tests/kit/model_zoo/transformers/chatglm.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import torch
+import transformers
+
```
Comment (FrankLeeeee) [inline tests/kit/model_zoo/transformers/chatglm.py:3]: Please follow the template instruction to make your PR more standard

File: tests/kit/model_zoo/transformers/chatglm.py (hunk 4 above)
Comment (klhhhhh) [reply tests/kit/model_zoo/transformers/chatglm.py:15]: they're not included the config class, which means you can't not pass them directly to the ChatGLMConfig, but they will be used when training stage

File: tests/kit/model_zoo/transformers/chatglm2_6b/MODEL_LICENSE (hunk 5 above)
Comment (klhhhhh) [reply tests/kit/model_zoo/transformers/chatglm2_6b/MODEL_LICENSE:1]: ok

File: tests/kit/model_zoo/transformers/chatglm2_6b/configuration_chatglm.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        self.bias_dropout_fusion = bias_dropout_fusion
+        self.multi_query_attention = multi_query_attention
+        self.multi_query_group_num = multi_query_group_num
+        self.apply_query_key_layer_scaling = apply_query_key_layer_scaling
+        self.attention_softmax_in_fp32 = attention_softmax_in_fp32
+        self.fp32_residual_connection = fp32_residual_connection
+        self.quantization_bit = quantization_bit
+        self.pre_seq_len = pre_seq_len
+        self.prefix_projection = prefix_projection
+        super().__init__(**kwargs)
```
Comment (FrankLeeeee) [inline tests/kit/model_zoo/transformers/chatglm2_6b/configuration_chatglm.py:59]: Please use pre-commit.

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**